pkg/libvmi/cloudinit.go

```diff
@@ -86,16 +86,33 @@ func addDiskVolumeWithCloudInitConfigDrive(vmi *v1.VirtualMachineInstance, diskN
 	addDisk(vmi, newDisk(diskName, bus))
 	v := newVolume(diskName)
 	v.VolumeSource = v1.VolumeSource{CloudInitConfigDrive: &v1.CloudInitConfigDriveSource{}}
-	addVolume(vmi, v)
+	addCloudInitVolume(vmi, v)
 }
 
 func addDiskVolumeWithCloudInitNoCloud(vmi *v1.VirtualMachineInstance, diskName string, bus v1.DiskBus) {
 	addDisk(vmi, newDisk(diskName, bus))
 	v := newVolume(diskName)
 	setCloudInitNoCloud(&v, &v1.CloudInitNoCloudSource{})
-	addVolume(vmi, v)
+	addCloudInitVolume(vmi, v)
 }
 
 func setCloudInitNoCloud(volume *v1.Volume, source *v1.CloudInitNoCloudSource) {
 	volume.VolumeSource = v1.VolumeSource{CloudInitNoCloud: source}
 }
+
+func addCloudInitVolume(vmi *v1.VirtualMachineInstance, volume v1.Volume) {
+	if !cloudInitVolumeExists(vmi, volume) {
+		vmi.Spec.Volumes = append(vmi.Spec.Volumes, volume)
+	}
+}
+
+func cloudInitVolumeExists(vmi *v1.VirtualMachineInstance, volume v1.Volume) bool {
+	for _, v := range vmi.Spec.Volumes {
+		if v.Name == volume.Name &&
+			(v.CloudInitNoCloud != nil && volume.CloudInitNoCloud != nil ||
+				v.CloudInitConfigDrive != nil && volume.CloudInitConfigDrive != nil) {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
The problem with introducing new Cirros factory outside `libfact` is that we will need to expose `cirrosMemory()`, which I recall Edy was against. 

The easiest way to get out of it is using Alpine VMs instead because they dont need the initial cloud init configuration (w/o it, boot take long time).
The resource footprint should be similar.

Reply:
Also, I recall @dankenigsberg suggested adding no-cloud-init option to libvmi to avoid this cloud-init volume mess.
Then it should look like so:
```
    libfact.NewCirros(
        libvmi.NoCloudInitNoCloud(),
        libvmi. WithCloudInitConfigDriveUserData()
    ...
```

WDYT?

Reply:
What is the meaning of `NoCloudInitNoCloud`?
I think that it will be great to have the default cloudinit in cirros only if no other cloudinit is specified in the options.
In this specific case, if we want to move this forward , we can use alpine

Reply:
The idea is to have `NoCloudInitNoCloud` remove existing cloudInitNoCloud volume, using it combined with `WithCloudInitConfigDriveUserData` will avoid the conflict

Reply:
Unfortunately its not possible to use alpine VMs because it works the same as cirros in a sense that `NewAlpine` factory adds cloud-init user-data volume by default to prevent long boot time :disappointed: 

Fallback to adding without-cloud-init-no-cloud `libvmi` option combined with using cloud-init config-drive `libvmi` options.
Please take another look :slightly_smiling_face: